src/green/builder.rs

```diff
@@ -1,23 +1,67 @@
-use rustc_hash::FxHashSet;
+use rustc_hash::{FxHashMap, FxHasher};
 
+use super::token::GreenTokenData;
 use crate::{
     green::{GreenElement, GreenNode, GreenToken, SyntaxKind},
     NodeOrToken, SmolStr,
 };
+use smallvec::SmallVec;
+use std::{
+    fmt::Debug,
+    hash::{Hash, Hasher},
+};
 
 #[derive(Default, Debug)]
 pub struct NodeCache {
-    nodes: FxHashSet<GreenNode>,
-    tokens: FxHashSet<GreenToken>,
+    nodes: FxHashMap<GreenNodeHash, GreenNode>,
+    tokens: FxHashMap<GreenTokenData, GreenToken>,
+}
+
+struct GreenNodeHash {
+    hasher: FxHasher,
+    inner_hash: u64,
+}
+
+impl Eq for GreenNodeHash {}
+impl PartialEq for GreenNodeHash {
+    fn eq(&self, other: &Self) -> bool {
+        self.inner_hash == other.inner_hash
+    }
+}
+
+impl Debug for GreenNodeHash {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.inner_hash.fmt(f)
+    }
+}
+
+impl GreenNodeHash {
+    fn new<'a>(kind: SyntaxKind) -> Self {
+        let mut hasher = FxHasher::default();
+        kind.hash(&mut hasher);
+        let inner_hash = hasher.finish();
+        Self { hasher, inner_hash }
+    }
+
+    fn add_child(&mut self, child: &GreenElement) {
+        child.hash(&mut self.hasher);
+        self.inner_hash = self.hasher.finish();
+    }
+}
+
+impl Hash for GreenNodeHash {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.inner_hash.hash(state);
+    }
 }
 
 impl NodeCache {
     fn node<I>(&mut self, kind: SyntaxKind, children: I) -> GreenNode
     where
-        I: IntoIterator<Item = GreenElement>,
-        I::IntoIter: ExactSizeIterator,
+        I: ExactSizeIterator<Item = GreenElement>,
```

Comment:
You can avoid this API change by just starting the method with a `let children = children.into_iter()`.